sql/api/src/main/scala/org/apache/spark/sql/types/StructField.scala

```diff
@@ -124,6 +124,13 @@ case class StructField(
       throw SparkException.internalError(s"Unexpected data type $dt")
   }
 
+  /**
+   * Updates the StructField with a new dataType.
+   */
+  def withNewType(dataType: DataType): StructField = {
```

Comment:
```suggestion
  private [sql] def withNewType(dataType: DataType): StructField = {
```
Or even something more restrictive please because this must not be called if equalsIgnoreCompatibleCollation does not hold. Do we maybe not put it here but in ddl.scala? I believe copy is public